### PR TITLE
doc/reference/storage/settings: storage behavior description

### PR DESCRIPTION
**Pull request by nvlsianpu**
Elaborated on storage mechanism usage and key-value version
garbage collection.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

File: doc/reference/storage/settings/settings.rst [hunk 1]
```diff
@@ -70,6 +70,23 @@ After all data is loaded, the ``h_commit`` handler is issued,
 signalling the application that the settings were successfully
 retrieved.
 
+Storing data to persisted storage
```
Comment (carlescufi) [inline doc/reference/storage/settings/settings.rst]: persistent

File: doc/reference/storage/settings/settings.rst [hunk 2]
```diff
@@ -70,6 +70,23 @@ After all data is loaded, the ``h_commit`` handler is issued,
...
+*********************************
+
+A call to ``settings_save_one()`` and ``settings_save()`` uses an ``h_export``
+implementation to store settings data to the storage medium.
+For both FCB and filesystem back-end only storage request with data which
```
Comment (carlescufi) [inline doc/reference/storage/settings/settings.rst]: requests

File: doc/reference/storage/settings/settings.rst [hunk 3]
```diff
@@ -70,6 +70,23 @@ After all data is loaded, the ``h_commit`` handler is issued,
...
+changes most actual key's value are stored, therefore there is no need to check
+whether a value changed by the application. Such storage mechanism implicate
```
Comment (carlescufi) [inline doc/reference/storage/settings/settings.rst]: implies

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: doc/reference/storage/settings/settings.rst [hunk 4]
```diff
@@ -70,6 +70,23 @@ After all data is loaded, the ``h_commit`` handler is issued,
...
+that storage can contain multiple of values for a key, while only the last is
+the current value for the key.
+
+Garbage collection
+==================
+When storage comes to be full (FCB) or consumes to much room (file system),
+the backend tries to remove some unneeded key-value pairs:
```
Comment (dbkinder) [inline doc/reference/storage/settings/settings.rst]: the backend removes non-recent key-value pairs and unneeded key-delete records.

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Comment by nvlsianpu** [comment]
updated

File: doc/reference/storage/settings/settings.rst [hunk 5]
```diff
@@ -70,6 +70,23 @@ After all data is loaded, the ``h_commit`` handler is issued,
 signalling the application that the settings were successfully
 retrieved.
 
+Storing data to persistent storage
+**********************************
+
+A call to ``settings_save_one()`` and ``settings_save()`` uses an ``h_export``
```
Comment (fnde-ot) [inline doc/reference/storage/settings/settings.rst]: AFAIK only **settings_save** uses the module's h_export. **settings_save_one** have both key and value passed as arguments.

Comment (nvlsianpu) [reply doc/reference/storage/settings/settings.rst]: Right - corrected.

**Review by fnde-ot — APPROVED** [review]
Looks good to me!

**Comment by nvlsianpu** [comment]
@dbkinder  @carlescufi  shippable failed on timer test which are unrelated to this patch

**Review by carlescufi — APPROVED** [review]

File: doc/reference/storage/settings/settings.rst [hunk 6]
```diff
@@ -70,6 +70,27 @@ After all data is loaded, the ``h_commit`` handler is issued,
 signalling the application that the settings were successfully
 retrieved.
 
+Storing data to persistent storage
+**********************************
+
+A call to ``settings_save_one()`` uses a backend implementation to store
+settings data to the storage medium. A call to ``settings_save()`` uses an
+``h_export`` implementation to store data bulky using ``settings_save_one()``.
```
Comment (dbkinder) [inline doc/reference/storage/settings/settings.rst]: ``store data bulky`` doesn't sound right. Did you mean ``store bulky data``?  But then, I'm not sure what you mean.

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Comment by nashif** [comment]
@nvlsianpu ping

File: doc/reference/storage/settings/settings.rst (hunk 6 above)
Comment (nvlsianpu) [reply doc/reference/storage/settings/settings.rst]: I wan't to siad that  lot of different data are stored in one "operation"

Comment (nashif) [reply doc/reference/storage/settings/settings.rst]: " store data bulky " does not sound right.

Comment (dbkinder) [reply doc/reference/storage/settings/settings.rst]: so, how about saying:
```
A call to ``settings_save()`` uses an
``h_export`` implementation to store different data in one
operation using ``settings_save_one()``.
```

**Review by dbkinder — DISMISSED** [review]

File: doc/reference/storage/settings/settings.rst (hunk 6 above)
Comment (nvlsianpu) [reply doc/reference/storage/settings/settings.rst]: @dbkinder Applied your suggestion.

**Comment by nvlsianpu** [comment]
@dbkinder Fixed - please revisit in your convince.

File: doc/reference/storage/settings/settings.rst [hunk 7]
```diff
@@ -70,6 +70,28 @@ After all data is loaded, the ``h_commit`` handler is issued,
...
+
+For both FCB and filesystem back-end only storage requests with data which
+changes most actual key's value are stored, therefore there is no need to check
+whether a value changed by the application. Such a storage mechanism implies
+that storage can contain multiple value assignments for a key , while only the
+last is the current value for the key.
+
+Garbage collection
+==================
+When storage becomes full (FCB) or consumes too much room (file system),
```
Comment (rakons) [inline doc/reference/storage/settings/settings.rst]: room -> space

File: doc/reference/storage/settings/settings.rst [hunk 8]
```diff
@@ -70,6 +70,28 @@ After all data is loaded, the ``h_commit`` handler is issued,
...
+the backend removes non-recent key-value pairs records and unneeded key-delete
```
Comment (rakons) [inline doc/reference/storage/settings/settings.rst]: unneeded -> unused ? Sounds better, but leaving it just as a comment.

Comment (rakons) [reply doc/reference/storage/settings/settings.rst]: Or even - unnecessary. It is just... Unneeded sounds more emotional, not technical. But maybe it is just me.

File: doc/reference/storage/settings/settings.rst (hunk 7 above)
Comment (nvlsianpu) [reply doc/reference/storage/settings/settings.rst]: done

File: doc/reference/storage/settings/settings.rst (hunk 8 above)
Comment (nvlsianpu) [reply doc/reference/storage/settings/settings.rst]: I found unused badly,  the `unnecessary` well reflects the intent of this poetry.

**Review by rakons — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@nashif ?

**Review by carlescufi — APPROVED** [review]

**Pull request merged**